### ISSUE TITLE
Skipped tests take a long time

### ISSUE THREAD

**Issue by volnoboy** [serenity-bdd/serenity-core#1016]
I am on Serenity version 1.6.6.

I have added to some tests in Jbehave metatags to skip:
Meta:
@skip
@ignored true

For some reason, some skipped steps take 0 seconds and some take few minutes. Web driver didn't even open.

Screenshot: <a href="https://ibb.co/mnTWQw"><img src="https://preview.ibb.co/dLxfJG/Skipped.png" alt="Skipped" border="0"></a>

**Comment by wakaleo** [comment]
Use @skip or @ignored, not both. Skipped tests will execute top-level code in your step definitions of it is not wrapped in @Step methods. Check for setup code or logic not wrapped in @Step methods

**Comment by wakaleo** [comment]
In short, this is most likely normal behaviour

**Comment by volnoboy** [comment]
1. @ignored true doesn't skip tests. Only @skip does.

2. I have investigated step that takes longer. It just calls the method that has @Step annotation. And it isn't heavy step, just doing few clicks. Without @skip it would take 20 seconds tops, with click takes 225 seconds.

**Comment by wakaleo** [comment]
Do a starch dump to see where it is slowing down. 

(“@ignored true” is not correct, did you read this notation somewhere?)

**Comment by volnoboy** [comment]
How to do a starch dump? 

Official documentation on Jbehave, it has both meta tags:
http://jbehave.org/reference/latest/meta-info.html

**Comment by wakaleo** [comment]
I think the examples in JBehave are just examples, there is no implicit behaviour of these tags for JBehave. Serenity uses @skip and @ignore. 

How you do a stack dump will depend on your IDE and/or OS.

**Comment by volnoboy** [comment]
First of all, you wrote @Skip with capital S and it does not work, only with small letter works.

I didn't figure out how to do stack dump, but I found the reason. It executes all code even within PageObject. 
I have a custom method to click multiple times and it is slow because of this method:
It is going inside catch clause and executes loop all 20 times. This is definitely should not be happening.

```
    protected void managedClick(WebElementFacade element) {
        for (int i = 0; i <= 20; i++) {
            try {
                element.click();
                break;
            } catch (Throwable e) {
                waitABit(248);
            }
        }
    }
```

**Comment by wakaleo** [comment]
Code that is not wrapped in a @Step method will be called normally. Serenity expects your step definition method to be delegating to @Step methods rather than calling UI logic directly. It looks like you are invoking this method directly from the JBehave step definitions, and not via a @Step-annotated method in a step library that represents your business tasks (which is the recommended layered architecture). 

Try upgrading to serenity-core 1.7.4, which has some safeguards in this area. 

On a side note, if you want to be sure that an element is present before clicking, a for loop with an explicit wait is not a very good way to do it. The WebElementFacade method will already wait for the element to be available (for up to 5 seconds, or whatever you configure in the `webdriver.wait.for.timeout` property) when you call the click() method.

**Comment by volnoboy** [comment]
I have 3 layers in my test application. 
1. Jbehave methods with @When annotation.
2. Serenity Steps layer, **where every method has @Step annotation**
3. Page Object layer.

This method, that takes too long, is in Page Object layer and is inside @Step annotated method.

Our application is angular 4, lots of ajax. Unfortunately, Serenity waits are not working for us.

I will try to update to  1.7.4 and see if it solves the issue.

**Comment by volnoboy** [comment]
Thanks, looks fixed in 1.7.4. But I get an exception when using openAt() method of PageObject class:

[pool-1-thread-1] ERROR net.serenitybdd.core.pages.PageObject - Failed to analyse default page URL: Starting URL: "_MY_URL_", Base URL:"_MY_BASE_URL_"
[pool-1-thread-1] ERROR net.serenitybdd.core.pages.PageObject - URL analysis failed with exception:
java.net.MalformedURLException: no protocol:  "_MY_URL_"
	at java.net.URL.<init>(URL.java:593)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at net.serenitybdd.core.pages.PageObject.replaceHost(PageObject.java:630)
	at net.serenitybdd.core.pages.PageObject.updateUrlWithBaseUrlIfDefined(PageObject.java:620)
	at net.serenitybdd.core.pages.PageObject.openAt(PageObject.java:784)

**Comment by wakaleo** [comment]
Can you post the code for your page object?

**Comment by volnoboy** [comment]
In my class that extends ScenarioSteps I have a method: 

```
@Step
    public void openFactoryDetailsPage(String id) {
        factoryDetailsPage.openAt("_my_url_");
    }
```

**Comment by volnoboy** [comment]
I just changed:
factoryDetailsPage.openAt("_my_url_");
to:
getDriver().get(_my_url_);

Looks like exception is gone

**Comment by wakaleo** [comment]
"MY_URL" and "MY_BASE_URL" might be defined in your code or serenity.properties file. Do you have a @DefaultUrl defined for this Page Object?

**Comment by volnoboy** [comment]
Only in serenity.properties file, this Page Object has no @defaulturl defined.

**Comment by wakaleo** [comment]
What is the content of your serenity.properties file?

**Comment by volnoboy** [comment]
webdriver.driver=chrome
webdriver.remote.driver=chrome
serenity.console.colors=true
webdriver.chrome.driver=chromedriver.exe
webdriver.ie.driver=IEDriverServer.exe
webdriver.edge.driver=MicrosoftWebDriver.exe
webdriver.base.url=MY_URL
serenity.project.name=CrediFi Automation
webdriver.timeouts.implicitlywait=5000
serenity.take.screenshots=AFTER_EACH_STEP
serenity.restart.browser.for.each=STORY
story.timeout.in.secs=18000
serenity.timeout=5000
ignore.failures.in.stories=true
chrome.switches=--disable-infobars;--ignore-certificate-errors;
json.pretty.printing=true

**Comment by wakaleo** [comment]
webdriver.base.url is the URL used when you open page objects, based on the default URL. You place the URL of each page in the @DefaultURL, you can use the open() method. If you use the openAt() method, the URL will be calculated relative to the base url. If you are not using the @DefaultURL annotation, getDriver().get(my_url) should be sufficient.

**Comment by volnoboy** [comment]
Ok, Thank you.

**Comment by wakaleo** [comment]
There was a scenario where WebDriver calls on Page Objects were called after a skipped test - this is fixed in 1.7.5
